### ISSUE TITLE
Introducing a spatial index interface

### ISSUE THREAD

**Issue by asinghvi17** [JuliaGeo/GeoInterface.jl#131]
There are multiple implementations of spatial index trees in Julia: SortTileRecursiveTree, SpatialIndexing, LibSpatialIndex.  The problem lies in that there is no single API to query any of these.  They all return the same thing, we just need a common API to query them.

If we get GeoPackage.jl going then there is even the possibility of extracting the SQLite R-tree from there, which would be interesting.  In any case, to my mind it makes sense to have some form of interface here - maybe the following on the implementor's side,
```julia
GI.trait(x)::GI.SpatialIndexTrait
GI.query(::GI.SpatialIndexTrait, x, extent)::Vector{Int}
```

Then, in GI,
```julia
GI.query(x, geom) = GI.query(GI.trait(x), GI.trait(geom), x, geom)
GI.query(::SpatialIndexTrait, ::AbstractGeometryTrait, x, geom) = GI.query(SpatialIndexTrait(), x, GI.extent(geom))
GI.query(::SpatialIndexTrait, ::Nothing, x, ext::Extent) = GI.query(SpatialIndexTrait(), x, ext)
```

**Comment by asinghvi17** [comment]
The join part is already implemented in [JuliaGeo/GeometryOps#75](https://github.com/JuliaGeo/GeometryOps.jl/pull/113) - we would just need a way to get an arbitrary index from, say, a database, or a pregenerated index.  LibSpatialIndex is also not threadsafe, so I would rather use SpatialIndexing.jl or RTrees.jl.

I think the lowest common denominator is that all of these spatial indexing methods return a list of neighbor indices given some area of interest, though?  For spatial joins and other similar tasks, getting a list of neighbours is all you need, since they have to be filtered by some predicate afterwards in any case.  

(Now that you mention it, perhaps knn trees may not follow this interface - but I haven't heard of those being used for GIS either...)

**Comment by asinghvi17** [comment]
Now that I think about it, we could implement this first in GeometryOps, to see if it makes sense, and then upstream to GeoInterface once we've hashed it out...package extensions in GeometryOps should do the work for this!

**Comment by rafaqz** [comment]
> Now that I think about it, we could implement this first in GeometryOps

Absolutely, playing with new ideas is pretty hard post 1.0. Even `GO.apply` could move to GeoInterface, but best to wait until its been stable for a long time before we do anything like that. For that reason I kind of regret putting the geometry types in GeoInterface rather than in a separate but closely related package.
